libcontainer/rootfs_linux.go

```diff
@@ -509,13 +552,85 @@ func mountToRootfs(c *mountConfig, m mountEntry) error {
 				return err
 			}
 		}
-		// bind mount won't change mount options, we need remount to make mount options effective.
-		// first check that we have non-default options required before attempting a remount
-		if m.Flags&^(unix.MS_REC|unix.MS_REMOUNT|unix.MS_BIND) != 0 {
-			// only remount if unique mount options are set
-			if err := remount(m, rootfs, c.noMountFallback); err != nil {
-				return err
+
+		// The initial MS_BIND won't change the mount options, we need to do a
+		// separate MS_BIND|MS_REMOUNT to apply the mount options.
+		if err := utils.WithProcfd(rootfs, m.Destination, func(dstFD string) error {
```

Comment:
I see the old code has a check `if m.Flags&^(unix.MS_REC|unix.MS_REMOUNT|unix.MS_BIND) != 0 {`, you removed it, so we will always do `remount` here?

Reply:
Maybe we should add a testcase for `rbind`?

Reply:
This is intentional -- we want to always do the remount to update the mount flags. If you look a the new tests, most of them would fail without this change (doing a "rw" bind-mount of an "ro" filesystem would silently produce an ro bind-mount with the old logic -- same goes for every other VFS flag).

Reply:
What kind of test were you thinking of?

Reply:
> doing a "rw" bind-mount of an "ro" filesystem would silently produce an ro bind-mount with the old logic -- same goes for every other VFS flag

But do you think this is right?

I have a testcase, the behavior has a difference between host and container:
In the host:
```
root@acmcoder:/home/acmcoder/mounttest# mkdir -p src/rbinddir rbinddir rodest rwdest
root@acmcoder:/home/acmcoder/mounttest# mount -o bind,rw rbinddir src/rbinddir
root@acmcoder:/home/acmcoder/mounttest# mount -o bind,ro src rodest
root@acmcoder:/home/acmcoder/mounttest# mount -o bind,rw rodest rwdest
root@acmcoder:/home/acmcoder/mounttest# echo test > rodest/abctest
bash: rodest/abctest: 只读文件系统(Read-only filesystem)
root@acmcoder:/home/acmcoder/mounttest# echo test > rwdest/abctest
bash: rwdest/abctest: 只读文件系统(Read-only filesystem)
root@acmcoder:/home/acmcoder/mounttest# echo test > rodest/rbinddir/abctest
bash: rodest/rbinddir/abctest: 只读文件系统(Read-only filesystem)
root@acmcoder:/home/acmcoder/mounttest# echo test > rwdest/rbinddir/abctest
bash: rwdest/rbinddir/abctest: 只读文件系统(Read-only filesystem)
root@acmcoder:/home/acmcoder/mounttest# echo test > src/rbinddir/abctest
root@acmcoder:/home/acmcoder/mounttest#
```

With the mount config `{"destination":"/opt/abc","type":"bind","source":"/home/acmcoder/mounttest/rodest","options":["rbind"]}` to run a container:
```
root@acmcoder:/home/acmcoder/ubuntutest# /home/acmcoder/go/src/github.com/opencontainers/runc/runc run test
# echo test > /opt/abc/abctest
# echo test > /opt/abc/rbinddir/abctest
# cat /opt/abc/abctest
test
#
```

Reply:
The old check didn't produce the behaviour of `mount(8)` either -- it produced really broken behaviour if you set one flag on a mount that had other flags set (it would only clear existing flags if you requested a new flag be set).

The issue is that it's unclear what semantics the runtime-spec wants -- it's not enough to say you want `mount(8)` behaviour. `mount --bind -o ...` and `mount --bind ; mount --bind -o remount` have different behaviour -- this PR implements as close as we can get to the former (with somewhat better atime semantics), you're testing the latter.

The reason why I made it so that we don't inherit the original mount flags is because a container configuration should produce consistent results. But I could be convinced either way. The runtime-spec doesn't specify which is correct because you can get `mount(8)` to do either behaviour. Not to mention `mount(8)` has atime bugs, and it seems unfortunate that we would have to be bug-compatible with `mount(8)` as well...

I go into more detail on the issues [here](https://github.com/opencontainers/runc/pull/3967#issuecomment-1674302353) -- I planned to open a runtime-spec issue but I opened one against util-linux first. Basically the spec doesn't tell us what to do here, and our current implementation is buggy (as is `mount(8)`, as it turns out). I think the behaviour I've picked is reasonable, but if we want to copy the `mount --bind -o remount,...` behaviour we can do that instead (but that would also be different to the existing behaviour -- the old check was completely broken in at least 3-4 different ways).

Reply:
Sorry for delete the last comment, I had some mistakes in it, but the result is the same.
Please see this, and don't need reply because you have replied in above comment.

> doing a "rw" bind-mount of an "ro" filesystem would silently produce an ro bind-mount with the old logic -- same goes for every other VFS flag
I think this is the right behavior. There is a case will result a different behavior between in host and in runc container:
In host:
```
root@acmcoder:/home/acmcoder/mounttest# mkdir -p rbinddir rodest rwdest src/rbinddir
root@acmcoder:/home/acmcoder/mounttest# mount -o bind rbinddir src/rbinddir
root@acmcoder:/home/acmcoder/mounttest# mount -o rbind,ro src rodest
root@acmcoder:/home/acmcoder/mounttest# mount -o rbind,rw rodest rwdest
root@acmcoder:/home/acmcoder/mounttest# echo foo > rodest/bar
bash: rodest/bar: Read-only Filesystem
root@acmcoder:/home/acmcoder/mounttest# echo foo > rwdest/bar
bash: rwdest/bar: Read-only Filesystem
root@acmcoder:/home/acmcoder/mounttest# echo foo > rwdest/rbinddir/bar
root@acmcoder:/home/acmcoder/mounttest# #mount -o remount,rw rwdest will make rwdest writeable
```

With a mount config `{"destination":"/opt/abc","type":"bind","source":"/home/acmcoder/mounttest/rodest","options":["rbind"]}` running a container:
```
root@acmcoder:/home/acmcoder/ubuntutest# ./runc run test
# echo foo1 > /opt/abc/bar
# echo foo2 > /opt/abc/rbinddir/bar
# cat /opt/abc/bar
foo1
#
root@acmcoder:/home/acmcoder/ubuntutest# cat ../mounttest/rodest/bar
foo1
```

Reply:
> it's not enough to say you want `mount(8)` behaviour. `mount --bind -o ...` and `mount --bind ; mount --bind -o remount` have different behaviour

Yes, but I think maybe mount option `remount` in `config.json` means we need use both, so maybe we should add a condition check `if m.Flags&unix.MS_REMOUNT != 0 {` before doing remount(but we should remove `remount` flag in the previous mount). WDTH.
I don't know the old check `unix.MS_REC` is for what, because estesp didn't add a testcase for that, so I said we may need to add a testcase for `rbind` or other.

Reply:
The problem is that we don't have a `remount` flag in the runtime-spec, and as a user I would expect a mount setting of `["bind", "rw"]` to be a read-write mount. The fact it isn't is a bug. In fact, `mount(8)` has the exact same bug we have here.

I explain the differences between these modes and the logic behind this decision in [this comment](https://github.com/opencontainers/runc/pull/3967#issuecomment-1674302353). This PR implements `mount --bind -o flags` but without the bug that clearing flags are handled incorrectly.

Here is an example of the `mount --bind -o flags` bug:

```
% mkdir /tmp/{foo,bar,baz}
% mount -t tmpfs tmpfs /tmp/foo
% mount --bind -o remount,ro /tmp/foo # set MS_RDONLY
% mount --bind -o rw /tmp/foo /tmp/bar # MS_RDONLY is not cleared because there are no new flag requested -- this is a bug!
% mount --bind -o noatime /tmp/foo /tmp/baz # MS_RDONLY is cleared because another flag was requested
```

I have [filed an issue with util-linux](https://github.com/util-linux/util-linux/issues/2433) about this. 

It seems to me that users expect `mount --bind -o flags` behaviour (only set the flags I actually asked for, clear the rest) but without the above bugs. If you want `--mount -o remount,...` behaviour (which appears to be what you're saying) then we need to get the old mount flags and copy them (only clearing flags the user explicitly requested be cleared). This is also _fine_ but is a much larger change to our old behaviour and I am not entirely sure users would expect that. If a user is bind-mounting a nosuid mount source, would they expect to have to specify `suid`? Maybe it would be preferable but I suspect people depend on the current behaviour and we can't change it.

Reply:
Maybe we could implement it so that we copy the old mount flags unless _any_ flag is specified, but that's making the semantics more complicated and I don't know if users would expect that behaviour either.

Reply:
> The problem is that we don't have a `remount` flag in the runtime-spec, and as a user I would expect a mount setting of `["bind", "rw"]` to be a read-write mount.

To mount a read-write mount, maybe need the config like this:  `["bind", "remount", "rw"]`, just like people use it with the `mount` command.
`["bind", "rw"]` means: `mount --bind -o rw /tmp/foo /tmp/bar`
`["bind", "remount", "rw"]` means: `mount --bind /tmp/foo /tmp/bar` and `mount -o remount,rw /tmp/bar`

But as you say, it looks very complicated.

Reply:
We could add a remount flag to the spec, but the issue is whether we should be bug-compatible with `mount(8)` when it's not specified? This implementation is _not_ bug compatible, but a bug compatible implementation has other issues...

If you pass `["bind", "rw", "nosuid"]` with current runc on a `ro,noexec` mount, you get a `rw,nosuid` mount (ro and noexec flags cleared). If you remove the nosuid from the config you get a `noexec,ro` mount (rw is silently ignored and all old flags are kept). I suspect no user expects this behaviour, even if it is somewhat similar to `mount(8)` (but not exactly).

Reply:
Mmm, I have got your thought now. There is an interesting thing: `remount` in `config.json` is legal, but will fail in anyway.

Reply:
> (doing a "rw" bind-mount of an "ro" filesystem would silently produce an ro bind-mount with the old logic -- same goes for every other VFS flag)

But if we doing a default bint-mount of an "ro" filesystem would silently produce an 'rw` bind-mount with your new logic. I think this is an new problem. So I'll take back my LGTM and until I or you have a time to see this problem.

Reply:
> doing a default bind-mount

The mount options is: ["bind"] or ["rbind"].

Reply:
Sorry for my long comments, somewhere in the previous comments I suggested that maybe we should default to keeping the flags if there are no other settings (*including* clearing flags like `rw`). I'll send an additional patch for you to review.

Reply:
Will the additional patch be in this PR or an new PR after this one got merged?

Reply:
The PR has been updated with this change.